File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorDeserializer/ResultRoller.java
package tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class ResultRoller
{
    private GeneratorCategory rootGeneratorCategory;
    private Random rng;

    public ResultRoller(GeneratorCategory rootGeneratorCategory)
    {
        this.rootGeneratorCategory = rootGeneratorCategory;
        this.rng = new Random();
    }

    public List<ResultItem> generateResultSet(String fullyQualifiedGeneratorPath, int numRolls)
    {
        return generateResultSet(fullyQualifiedGeneratorPath, numRolls, 1, Integer.MAX_VALUE - 1);
    }

    public List<ResultItem> generateResultSet(String fullyQualifiedGeneratorPath, int numRolls, int minRollValue, int maxRollValue)
    {
        List<ResultItem> resultSet = new ArrayList<ResultItem>();

        Generator randomTable = rootGeneratorCategory.getGeneratorFromFullPath(fullyQualifiedGeneratorPath, rootGeneratorCategory);
        int tableMaxRollValue = determineTableDiceRange(randomTable.getTable());

        for (int i = 0; i < numRolls; ++i) {
            int actualMinRoll = minRollValue > tableMaxRollValue ? tableMaxRollValue : minRollValue;
            int actualMaxRoll = tableMaxRollValue < maxRollValue ? tableMaxRollValue : maxRollValue;
            int roll = rng.nextInt((actualMaxRoll - actualMinRoll) + 1) + minRollValue;

            for (TableEntries tableEntry : randomTable.getTable()) {
                boolean entryIsMatch = checkRollResult(roll, tableEntry);

                if (entryIsMatch) {
                    Map<String, String> rerollSubTable = tableEntry.getRerollSubTable();

                    //Some entries on random tables call for a roll on a completely different table
                    //So handle that here with a recursive call + merging the resulting sets of
                    //ResultItems together
                    if (rerollSubTable != null && rerollSubTable.isEmpty() == false) {
                        String subTableName = rerollSubTable.get("SubTableName");
                        SubTableRollRange rollRange = tableEntry.getSubTableRollRange();
                        int numSubTableRolls = tableEntry.getNumSubTableRolls();

                        List<List<ResultItem>> subTableResults = new ArrayList<List<ResultItem>>();
                        for (int rollIteration = 0; rollIteration < numSubTableRolls; ++rollIteration) {
                            subTableResults.add(
                                generateResultSet(subTableName, numSubTableRolls, rollRange.getMinRoll(), rollRange.getMaxRoll())
                            );
                        }

                        subTableResults.add(resultSet);
                        resultSet = mergeResultSets(subTableResults);

                        continue;
                    }

                    //Now the simple case. Add the new item directly, or increment an existing value.
                    ResultItem newItem = new ResultItem(fullyQualifiedGeneratorPath, tableEntry.getName(), tableEntry.getMetadata());
                    resultSet = addResultItem(resultSet, newItem);
                }
            }
        }

        return resultSet;
    }

    private List<ResultItem> addResultItem(List<ResultItem> sourceList, ResultItem newItem)
    {
        List<ResultItem> resultList = sourceList;

        for (ResultItem item : resultList) {
            if (item.getName().equals(newItem.getName())) {
                item.setQuantity(item.getQuantity() + 1);
                return resultList;
            }
        }

        resultList.add(newItem);
        return resultList;
    }

    private List<ResultItem> mergeResultSets(List<List<ResultItem>> mergeSets)
    {
        List<ResultItem> finalMergedSet = new ArrayList<ResultItem>();

        for (List<ResultItem> mergeSet : mergeSets) {
            for (ResultItem item : mergeSet) {

                boolean hadMatchInFinalMergedSet = false;
                for (ResultItem mergedItem : finalMergedSet) {
                    if (item.getName().equals(mergedItem.getName())) {
                        hadMatchInFinalMergedSet = true;
                        mergedItem.setQuantity(mergedItem.getQuantity() + 1);
                        break;
                    }
                }

                if (hadMatchInFinalMergedSet == false) {
                    finalMergedSet.add(item);
                }
            }
        }

        return finalMergedSet;
    }

    private boolean checkRollResult(int roll, TableEntries tableEntry)
    {
        String diceRange = tableEntry.getDiceRange();
        int rangeSeparatorIndex = diceRange.indexOf("-");

        //Entry contains a single value
        if (rangeSeparatorIndex == -1) {
            int targetValue = Integer.parseInt(diceRange);

            if (roll == targetValue) {
                return true;
            }

            return false;
        }

        //Entry contains a range of values. check if roll is between the range
        //inclusive on both ends
        int leftVal = Integer.parseInt(diceRange.substring(0, rangeSeparatorIndex));
        int rightVal = Integer.parseInt(diceRange.substring(rangeSeparatorIndex + 1));

        if (roll >= leftVal && roll <= rightVal) {
            return true;
        }

        return false;
    }

    private int determineTableDiceRange(TableEntries[] tableItems)
    {
        int tableMaxRoll = 1;

        for (TableEntries item : tableItems) {
            String diceRange = item.getDiceRange();
            int rangeSeparatorIndex = diceRange.indexOf("-");

            //Entry contains a single value
            if (rangeSeparatorIndex == -1) {
                int rollValue = Integer.parseInt(diceRange);

                if (rollValue > tableMaxRoll) {
                    tableMaxRoll = rollValue;
                }

                continue;
            }

            //Entry contains a range of values. Two integers hyphen-separated
            int leftVal = Integer.parseInt(diceRange.substring(0, rangeSeparatorIndex));
            int rightVal = Integer.parseInt(diceRange.substring(rangeSeparatorIndex + 1));

            if (leftVal > tableMaxRoll) {
                tableMaxRoll = leftVal;
            }

            if (rightVal > tableMaxRoll) {
                tableMaxRoll = rightVal;
            }
        }

        return tableMaxRoll;
    }
}

File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorDeserializer/ResultRoller.kt
package tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer

import tabletop.velocic.com.worldforgerpgtools.Extensions.random

class ResultRoller(
    private val rootGeneratorCategory: GeneratorCategory
) {
    fun generateResultSet(fullyQualifiedGeneratorPath: String, numRolls: Int = 1) : List<ResultItem> {
        return generateResultSet(fullyQualifiedGeneratorPath, numRolls, 1, Int.MAX_VALUE - 1)
    }

    fun generateResultSet(fullyQualifiedGeneratorPath: String, numRolls: Int = 1, minRollValue: Int, maxRollValue: Int) : List<ResultItem> {
        var resultSet: MutableList<ResultItem> = mutableListOf()

        val randomTable = rootGeneratorCategory.getGeneratorFromFullPath(fullyQualifiedGeneratorPath, rootGeneratorCategory).table
        val tableRollRange = determineTableDiceRange(randomTable)

        for (i in 0 until numRolls) {
            val actualMinRoll = if (minRollValue > tableRollRange.last) tableRollRange.last else minRollValue
            val actualMaxRoll = if (tableRollRange.last < maxRollValue) tableRollRange.last else maxRollValue

            val roll = (actualMinRoll..actualMaxRoll).random()

            for (entry in randomTable) {
                val entryIsMatch = roll in entry.diceRange

                if (entryIsMatch) {
                    val rerollSubTable = entry.rerollSubTable ?: mapOf()

                    //Some entries on random tables call for a roll on a completely different table
                    //So handle that here with a recursive call + merging the resulting sets of ResultItems
                    //together
                    if (!rerollSubTable.isEmpty()) {
                        val subTableName = rerollSubTable["SubTableName"] ?: ""
                        val subTableRollRange = entry.getSubTableRollRange()
                        val numSubTableRolls = entry.getNumSubTableRolls()
                        val subTableResults = mutableListOf(mutableListOf<ResultItem>())

                        for (subTableRoll in 0 until numSubTableRolls) {
                            subTableResults.add(generateResultSet(subTableName, numSubTableRolls, subTableRollRange.first, subTableRollRange.last).toMutableList())
                        }

                        subTableResults.add(resultSet)
                        resultSet = mergeResultSets(subTableResults)

                        continue
                    }

                    //Now the simple case. Add the new item directly, or increment an existing value
                    val newItem = ResultItem(fullyQualifiedGeneratorPath, entry.name, entry.metadata)
                    addResultItem(resultSet, newItem)
                }
            }
        }

        return resultSet
    }

    private fun mergeResultSets(mergeSets: MutableList<MutableList<ResultItem>>) : MutableList<ResultItem> {
        val finalMergedSet: MutableList<ResultItem> = mutableListOf()

        for (mergeSet in mergeSets) {
            for (item in mergeSet) {
                var hadMatchInFinalMergedSet = false

                for (mergedItem in finalMergedSet) {
                    if (item.name == mergedItem.name) {
                        hadMatchInFinalMergedSet = true
                        mergedItem.quantity = mergedItem.quantity + 1
                        break
                    }
                }

                if (!hadMatchInFinalMergedSet) {
                    finalMergedSet.add(item)
                }
            }
        }

        return finalMergedSet
    }

    private fun addResultItem(sourceList: MutableList<ResultItem>, newItem: ResultItem) : MutableList<ResultItem> {
        for (item in sourceList) {
            if (item.name == newItem.name) {
                item.quantity = item.quantity + 1
                return sourceList
            }
        }

        sourceList.add(newItem)
        return sourceList
    }

    private fun determineTableDiceRange(tableEntries: Array<TableEntries>): IntRange {
        var tableMinRoll = 1
        var tableMaxRoll = 1

        tableEntries.forEach {
            val entryRange = it.diceRange
            if (entryRange.first < tableMinRoll) {
                tableMinRoll = entryRange.first
            }
            if (entryRange.last > tableMaxRoll) {
                tableMaxRoll = entryRange.last
            }
        }

        return tableMinRoll..tableMaxRoll
    }
}